src/cmd.rs

```diff
@@ -433,25 +436,134 @@ impl fmt::Display for Value {
 
 ////////////////////////////////////////////////////////////////////////////////
 
+#[derive(Debug, PartialEq, Eq)]
+pub struct History {
+    /// History path.
+    pub path: std::path::PathBuf,
+    /// The history of commands entered.
+    entries: VecDeque<String>,
+    /// The current cursor into the history.
+    cursor: Option<usize>,
+    /// Maximum number of entries.
+    capacity: usize,
+}
+
+impl History {
+    pub fn new<P: AsRef<Path>>(path: P, capacity: usize) -> Self {
+        Self {
+            entries: VecDeque::new(),
+            cursor: None,
+            capacity,
+            path: path.as_ref().into(),
+        }
+    }
+
+    pub fn load(&mut self) -> io::Result<()> {
+        use std::fs::File;
+        use std::io::{BufRead, BufReader};
+
+        let f = File::open(&self.path)?;
+        let r = BufReader::new(f);
+
+        for line in r.lines() {
+            self.add(line?);
+        }
+        Ok(())
+    }
+
+    pub fn save(&self) -> io::Result<()> {
+        use std::fs::File;
+        use std::io::{BufWriter, Write};
+
+        if self.is_empty() {
+            return Ok(());
+        }
+
+        let parent = self
+            .path
+            .parent()
+            .expect("saving to a path with a parent directory");
```

Comment:
Is this non-recoverable at this point and therefore worth panicing?

Reply:
Yeah, the condition in which this happens is if `/` is passed in as the history file :joy: